File: Backend/Backend/Web_Analyzer/color/clr.py
import cssutils
from colour import Color

class CssColorExtractor:
    def __init__(self):
        self.properties_with_colors = [
            'color', 'background', 'background-color', 'background-image', 'border', 'border-top',
            'border-right', 'border-bottom', 'border-left', 'border-color', 'border-top-color',
            'border-right-color', 'border-bottom-color', 'border-left-color', 'outline',
            'outline-color', 'text-decoration', 'text-decoration-color', 'text-shadow',
            'box-shadow', 'fill', 'stroke', 'stop-color', 'flood-color', 'lighting-color',
        ]
        self.color_formats = [
            'hex', 'rgb', 'hsl', 'hwb', 'name',
        ]

    def does_property_allow_color(self, property):
        return property in self.properties_with_colors

    def is_color_grey(self, color):
        red = color.red * 255
        return self.is_color_monochrome(color) and 0 < red < 255

    def is_color_monochrome(self, color):
        return color.hsl[1] == 0

    def is_valid_color_format(self, format):
        return format in self.color_formats

    def serialize_color(self, value, options):
        if not options.get('color_format') or not self.is_valid_color_format(options['color_format']):
            return value

        color = Color(value)
        color_format = options['color_format']
        
        if color_format == 'name' and color.get_name():
            return color.get_name()
        
        if hasattr(color, color_format):
            formatted = getattr(color, color_format)
            if callable(formatted):
                return formatted()
            return formatted
        return value

    def default_options(self, options):
        return {
            'without_grey': False,
            'without_monochrome': False,
            'all_colors': False,
            'color_format': None,
            'sort': None,
            **options,
        }

    def sort_colors(self, colors, sort_options):
        options = self.default_options(sort_options)
        colors = [self.serialize_color(color, options) for color in colors]
        
        if options['sort'] == 'hue':
            colors.sort(key=lambda c: Color(c).hue)
        
        if options['sort'] == 'frequency':
            from collections import Counter
            frequency = Counter(colors)
            colors.sort(key=lambda x: -frequency[x])
        
        return colors if options['all_colors'] else list(set(colors))

    def extract_colors_from_string(self, string, options):
        colors = []
        options = self.default_options(options)

        import re
        items = re.split(r',\s*', string)
        for item in items:
            subitems = item.split()
            for subitem in subitems:
                match = re.match(r'^(?:-webkit-|-moz-|-o-)?(?:repeating-)?(?:radial|linear)-gradient\((.*?)\)$', subitem)
                if match:
                    colors.extend(re.split(r',\s*', match.group(1)))
                else:
                    colors.append(subitem)

        def is_valid_color(value):
            try:
                color = Color(value)
                if options['without_monochrome'] and self.is_color_monochrome(color):
                    return False
                if options['without_grey'] and self.is_color_grey(color):
                    return False
                if options['color_format'] == 'name' and not color.get_name():
                    return False
                return True
            except ValueError:
                return False

        return [color for color in colors if is_valid_color(color)]

    def extract_colors_from_decl(self, decl, options):
        if not self.does_property_allow_color(decl.name):
            return []
        return self.extract_colors_from_string(decl.value, options)

    def extract_colors_from_css(self, css, options):
        colors = []
        sheet = cssutils.parseString(css)

        for rule in sheet:
            if rule.type == rule.STYLE_RULE:
                for property in rule.style:
                    colors.extend(self.extract_colors_from_decl(property, options))

        return colors

    def from_decl(self, decl, options={}):
        return self.sort_colors(self.extract_colors_from_decl(decl, options), options)

    def from_css(self, css, options={}):
        return self.sort_colors(self.extract_colors_from_css(css, options), options)

    def from_string(self, string, options={}):
        return self.sort_colors(self.extract_colors_from_string(string, options), options)


# Example usage
if __name__ == "__main__":
    extractor = CssColorExtractor()

    css = """
    .example {
        color: #ff0000;
        background: linear-gradient(to right, #ff0000, #00ff00);
        border: 1px solid rgb(0, 0, 255);
    }
    """

    options = {
        'without_grey': False,
        'without_monochrome': False,
        'all_colors': True,
        'color_format': 'hex',
        'sort': 'frequency',
    }

    extracted_colors = extractor.from_css(css, options)
    print(extracted_colors)

File: Backend/Backend/Web_Analyzer/color/clr.js
'use strict';

const postcss = require('postcss');
const Color = require('color');

function CssColorExtractor() {
  const propertiesWithColors = [
    'color',
    'background',
    'background-color',
    'background-image',
    'border',
    'border-top',
    'border-right',
    'border-bottom',
    'border-left',
    'border-color',
    'border-top-color',
    'border-right-color',
    'border-bottom-color',
    'border-left-color',
    'outline',
    'outline-color',
    'text-decoration',
    'text-decoration-color',
    'text-shadow',
    'box-shadow',
    'fill',
    'stroke',
    'stop-color',
    'flood-color',
    'lighting-color',
  ];
  const colorFormats = [
    'hexString',
    'hexaString',
    'rgbString',
    'percentString',
    'hslString',
    'hwbString',
    'keyword',
  ];

  function doesPropertyAllowColor(property) {
    return propertiesWithColors.indexOf(property) > -1;
  }

  function isColorGrey(color) {
    const red = color.red();

    return isColorMonochrome(color) && red > 0 && red < 255;
  }


  function isColorMonochrome(color) {
    const hsl = color.hsl().object();

    return hsl.h === 0 && hsl.s === 0;
  }


  function isValidColorFormat(format) {
    return colorFormats.indexOf(format) > -1;
  }


  function serializeColor(value, options) {
    if (!options.colorFormat || !isValidColorFormat(options.colorFormat)) {
      return value;
    }
    const color = new Color(value);
    let colorFormat = options.colorFormat;
    if (!color[options.colorFormat]) {
      colorFormat = colorFormat.replace(/String$/, '');
    }
    const formatted = color[colorFormat]();
    if (typeof formatted === 'string') {
      return formatted;
    }
    return formatted.string();
  }


  function defaultOptions(options) {
    return Object.assign({
      withoutGrey: false,
      withoutMonochrome: false,
      allColors: false,
      colorFormat: null,
      sort: null,
    }, options);
  }


  function sortColors(colors, sortOptions) {
    const options = defaultOptions(sortOptions);
    colors = colors.map((value) => serializeColor(value, options));
    if (options.sort === 'hue') {
      colors = colors.sort((a, b) => {
        return new Color(a).hue() - new Color(b).hue();
      });
    }
    if (options.sort === 'frequency') {
      const frequencyMap = new Map();
      colors.forEach((value) => {
        frequencyMap.set(value, (frequencyMap.get(value) || 0) + 1);
      });
      colors = colors.sort((a, b) => {
        return frequencyMap.get(b) - frequencyMap.get(a);
      });
    }
    return options.allColors ? colors : unique(colors);
  }


  function unique(items) {
    return [...new Set(items)];
  }

  function extractColorsFromString(string, options) {
    let colors = [];

    const {
      withoutMonochrome,
      withoutGrey,
      colorFormat,
    } = defaultOptions(options);

    postcss.list.comma(string).forEach(function(items) {
      postcss.list.space(items).forEach(function(item) {
        const regex = new RegExp(
          '^' +
                    '(-webkit-|-moz-|-o-)?' +
                    '(repeating-)?' +
                    '(radial|linear)-gradient\\((.*?)\\)' +
                    '$',
        );

        const match = item.match(regex);

        if (match) {
          colors = colors.concat.apply(
            colors,
            postcss.list.comma(match[4]).map(postcss.list.space),
          );
        } else {
          colors.push(item);
        }
      });
    });

    return colors.filter((value) => {
      let color;
      try {
        color = new Color(value);
      } catch (e) {
        return false;
      }

      if (withoutMonochrome && isColorMonochrome(color)) {
        return false;
      }

      if (withoutGrey && isColorGrey(color)) {
        return false;
      }

      if (colorFormat === 'keyword') {
        const keywordColor = new Color(color.keyword());
        if (keywordColor.rgbNumber() !== color.rgbNumber()) {
          return false;
        }
      }

      return true;
    });
  }

  function extractColorsFromDecl(decl, options) {
    if (!doesPropertyAllowColor(decl.prop)) {
      return [];
    }

    return extractColorsFromString(decl.value, options);
  }

  function extractColorsFromCss(css, options) {
    let colors = [];

    postcss.parse(css).walkDecls(function(decl) {
      colors = colors.concat(extractColorsFromDecl(decl, options));
    });

    return colors;
  }

  this.fromDecl = (decl, options) => sortColors(extractColorsFromDecl(decl, options), options);
  this.fromCss = (css, options) => sortColors(extractColorsFromCss(css, options), options);
  this.fromString = (string, options) => sortColors(extractColorsFromString(string, options), options);
}

module.exports = new CssColorExtractor();